### PR TITLE
DataTrails: Exploring alternatives to history issues

### PR DESCRIPTION
**Pull request by torkelo**
* [x] Fixes issues with time range changes modifying the history step (https://github.com/grafana/grafana/pull/86741)
* [x] Fixes issues with filters modifying a history step  (https://github.com/grafana/grafana/pull/86817) 

Both were easy to fix as we did not clone the state when we restored state from a history state but used it directly. 

* [x] Tried fixing missing URL sync issues (selected tab no longer synced to URL) by restoring URL sync. Can't recall the reasons for only doing manual URL sync when switching history step.
* [x] Some minor code cleanup 

Still the major problem that is annoying me is that we create history step before queries complete, so the history step is really missing the "lastest state". But because the state changed event happen after the state change has occurred we cannot really use the latest state to create the previous step (which would be ideal as it would then include all the query results). Not sure how to fix that yet. 

TODO
* [x] Update unit tests,  could need help with this

### CODE REVIEW

File: public/app/features/trails/DataTrail.tsx [hunk 1]
```diff
@@ -170,25 +155,21 @@ export class DataTrail extends SceneObjectBase<DataTrailState> {
     return this.getMetricMetadata(this.state.metric);
   }
 
-  private goBackToStep(step: DataTrailHistoryStep) {
-    if (!step.trailState.metric) {
-      step.trailState.metric = undefined;
-    }
-
-    this.setState(step.trailState);
```
Comment (torkelo) [inline public/app/features/trails/DataTrail.tsx:178]: this the buggy code where just use the history step state directly

**Comment by darrenjaneczek** [comment]
>Still the major problem that is annoying me is that we create history step before queries complete, so the history step is really missing the "lastest state".

Can we create new events/subscribables in scenes that could be activated just before a variable change and/or state change takes effect?
Then maybe we could use these to store the step with the old state before the new state is applied onto the trail.

**Comment by darrenjaneczek** [comment]
> Can't recall the reasons for only doing manual URL sync when switching history step.

I will have to review. I'll search for any remaining side effects which may assume URL sync is manual.
These were related to back/forward browser button behaviors.

**Comment by darrenjaneczek** [comment]
> Update unit tests, could need help with this

I [incorporated unit tests](https://github.com/grafana/grafana/pull/86843/commits/cfa6e95d9dc48151bfd91d710cca6ec70b41daa4)  from:
 - https://github.com/grafana/grafana/pull/86817 
 - https://github.com/grafana/grafana/pull/86741

**Comment by darrenjaneczek** [comment]
> Still the major problem that is annoying me is that we create history step before queries complete, so the history step is really missing the "lastest state".

Could we listen for data query fetch completion and have the listeners insert updates to the history state?

Maybe rewrite the history state after the user leaves a step?

**Comment by torkelo** [comment]
@darrenjaneczek 
> Could we listen for data query fetch completion and have the listeners insert updates to the history state?
> Maybe rewrite the history state after the user leaves a step?

Yes, something like this, I think both ideas are worth persuing

**Review by darrenjaneczek — APPROVED** [review]
This PR is working perfectly. It satisfies the new test cases.
I also verified that this return to using URL sync is perfectly fine.
Thanks for your help on this!

**Comment by torkelo** [comment]
@darrenjaneczek no worries, feel partly to blame for making scenes so tricky 😢

**Comment by darrenjaneczek** [comment]
No worries as all. The capabilities it provides is worth climbing the learning curve and the occasional instructional hiccup.

**Pull request merged**